### PR TITLE
Frame simulator

### PR DESCRIPTION
**Pull request by skyfrench**
Initial implementation of a 'frame simulator' base plugin class / application for consideration; pull request also opened in dls-controls/excalibur-detector with example implementation.

### CODE REVIEW

**Comment by ulrikpedersen** [comment]
ah, it looks like there are some commits from the 'dls-master' branch in this PR. Those commits are DLS-specific and shouldn't get merged in here (I think). 

@skyfrench: I guess that's what you said yesterday about having to figure out what branch to merge to/from...? I'm not completely sure what is the easiest way to resolve this. Can you rebase your branch with your main changeset (i.e. those latest 7 commits) on top of odin-detector/master?

**Comment by ulrikpedersen** [comment]
The travis build seems to fail due to cmake not finding pcap. Try adding package "libpcap-dev" to the list in .travis.yml

File: frameSimulator/src/FrameSimulatorPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    }
+
+    /** Setup frame simulator plugin class from store of command line options
+     *
+     * \param[in] store of given command line options
+     * \param[out] true (false) if required program options are (not) specified
+     * and the simulator plugin is not ready to use
+     *
+     * Drived plugins must additionally call this base method even if overridden
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorPlugin.cpp]: Typo: Derived

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    int pcapFrameSimulatorPlugin::send_packet(const Packet& packet, const int& frame) const {
+        if (frame != curr_frame) {
+            curr_port_index = (curr_port_index + 1 < m_addrs.size()) ? curr_port_index + 1 : 0;
+            curr_frame = frame;
+        }
+        bind(m_socket, (struct sockaddr *) (&m_addrs[curr_port_index]), sizeof(m_addrs[curr_port_index]));
+        return sendto(m_socket, packet.data, packet.size, 0, (struct sockaddr *) (&m_addrs[curr_port_index]), sizeof(m_addrs[curr_port_index]));
+    }
+
+    void pcapFrameSimulatorPlugin::populate_options(po::options_description& config) {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs function comment

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        opt_destip.add_option_to(config);
+        opt_pcapfile.add_option_to(config);
+        opt_packetgap.add_option_to(config);
+        opt_dropfrac.add_option_to(config);
+        opt_droppackets.add_option_to(config);
+
+    }
+
+    pcapFrameSimulatorPlugin::~pcapFrameSimulatorPlugin() {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs function comment

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    /** Packet callback function
+     *
+     */
+    void pcapFrameSimulatorPlugin::pkt_callback(u_char *user, const pcap_pkthdr *hdr, const u_char *buffer) {
+        pcapFrameSimulatorPlugin* replayer = reinterpret_cast<pcapFrameSimulatorPlugin*>(user);
+        replayer->prepare_packets(hdr, buffer);
+    }
+
+    void pcapFrameSimulatorPlugin::simulate() {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs function comment

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    }
+
+    pcapFrameSimulatorPlugin::~pcapFrameSimulatorPlugin() {
+
+        close(m_socket);
+
+    }
+
+    /** Packet callback function
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs arguments in function comment

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+static const std::string librarySuffix = "FrameSimulatorPlugin";
+
+// Basic command line options
+static const FrameSimulatorOption<std::string> opt_detector("detector", "Set the detector (Excalibur, Eiger etc.)");
+static const FrameSimulatorOption<std::string> opt_libpath("lib-path", "Path to detector plugin library");
+static const FrameSimulatorOption<unsigned int> opt_debuglevel("debug-level", "Set the debug level");
+static const FrameSimulatorOption<std::string> opt_logconfig("logconfig", "Set the log4cxx logging configuration file");
+
+/** Load requested plugin
+ * /param[in] map of command line specified variables
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: I think the param comment should have the name of the variable (e.g. "vm"). [This comment also applies to other instances where the variable is not named in the comment in this PR]

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+ * the 'detector' argument must match the library name prefix i.e. 'lib<detector>FrameSimulatorPlugin.so'
+ */
+boost::shared_ptr<FrameSimulator::FrameSimulatorPlugin> get_requested_plugin(const po::variables_map& vm) {
+
+    std::string pluginClass = opt_detector.get_val(vm) + librarySuffix;
+
+    boost::filesystem::path libraryPathAndName = boost::filesystem::path(opt_libpath.get_val(vm)) /
+        boost::filesystem::path("lib" + pluginClass + ".so");
+
+boost::shared_ptr<FrameSimulator::FrameSimulatorPlugin> plugin = \
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: indentation

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    OdinData::ClassLoader<FrameSimulator::FrameSimulatorPlugin>::load_class(pluginClass, libraryPathAndName.string());
+
+    return plugin;
+
+}
+
+static bool has_suffix(const std::string &str, const std::string &suffix)
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: Needs function comment

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+{
+    return str.size() >= suffix.size() &&
+           str.compare(str.size() - suffix.size(), suffix.size(), suffix) == 0;
+}
+
+int parse_arguments(int argc, char** argv, po::variables_map& vm, LoggerPtr& logger,
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: Needs function comment

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            po::options_description config("Configuration options");
+
+            boost::filesystem::path libraryPathAndName = boost::filesystem::path(opt_libpath.get_val(vm)) /
+                                                         boost::filesystem::path("lib"+detector+librarySuffix+".so");
+
+            std::string pluginClass = detector + librarySuffix;
+
+            plugin = get_requested_plugin(vm);
+
+            if (plugin)
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: Think this should have { } even though it's only one line within the if.

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    LoggerPtr logger(Logger::getLogger("FS.App"));
+
+    try {
+
+        boost::shared_ptr<FrameSimulator::FrameSimulatorPlugin> plugin;
+
+        po::variables_map vm;
+        parse_arguments(argc, argv, vm, logger, plugin);
+
+        if (!plugin)
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: Think this should have { }

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+            LOG4CXX_ERROR(logger, "Unable to create simulator plugin, application will terminate");
+
+        // Setup plugin from command line arguments
+        if (plugin->setup(vm))
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: { }

File: frameSimulator/src/FrameSimulatorOption.cpp [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+namespace FrameSimulator {
+
+    /** Helper functions to parse comma separated required and optional command line arguments into lists
+     * /param[in] string of comma separated values
+     * /param[out] required output list as std::vector
+     */
+      void set_list_option(const std::string& option_val, std::vector<std::string>& list) {
+          boost::split(list, option_val, boost::is_any_of(","), boost::token_compress_on);
+      }
+      void set_optionallist_option(const std::string& option_val, boost::optional<std::vector<std::string> >& list) {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorOption.cpp]: Should have separate function comments for each function

File: frameSimulator/src/FrameSimulatorPlugin.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        opt_ports.add_option_to(config);
+        opt_frames.add_option_to(config);
+        opt_interval.add_option_to(config);
+
+    }
+
+    /** Setup frame simulator plugin class from store of command line options
+     *
+     * \param[in] store of given command line options
+     * \param[out] true (false) if required program options are (not) specified
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorPlugin.cpp]: Should be "return" not "param"

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+#include <net/ethernet.h>
+#include <netinet/udp.h>
+#include <netinet/ip.h>
+
+#include "pcapFrameSimulatorOptions.h"
+
+namespace FrameSimulator {
+
+    pcapFrameSimulatorPlugin::pcapFrameSimulatorPlugin() : FrameSimulatorPlugin() {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs function comment

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        // Setup logging for the class
+        logger_ = Logger::getLogger("FS.pcapFrameSimulatorPlugin");
+        logger_->setLevel(Level::getAll());
+
+        curr_frame = 0;
+        curr_port_index = 0;
+
+    }
+
+    bool pcapFrameSimulatorPlugin::setup(const po::variables_map& vm) {
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: Needs function comment

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        if (!opt_pcapfile.is_specified(vm)) {
+            LOG4CXX_ERROR(logger_, "pcap file is not specified");
+            return false;
+        }
+
+        // Optional arguments
+        opt_packetgap.get_val(vm, packet_gap);
+        opt_dropfrac.get_val(vm, drop_frac);
+
+        if (opt_droppackets.is_specified(vm))
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: { }

File: cmake/FindPCAP.cmake [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+include(CheckCSourceCompiles)
+set(CMAKE_REQUIRED_LIBRARIES ${PCAP_LIBRARY})
+check_c_source_compiles("int main() { return 0; }" PCAP_LINKS)
+set(CMAKE_REQUIRED_LIBRARIES)
+
+mark_as_advanced(
+    PCAP_ROOT_DIR
+    PCAP_INCLUDE_DIR
+    PCAP_LIBRARY
+)
```
Comment (MattTaylorDLS) [inline cmake/FindPCAP.cmake]: No new line at end of file

File: frameSimulator/include/FrameSimulatorOptions.h [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    /** Global / common frame simulator options
+     * used by all plugins
+     */
+    static const FrameSimulatorOption<std::string> opt_ports("ports", "Provide a (comma separated) port list", "9999");
+    static const FrameSimulatorOption<int> opt_frames("frames", "Number of frames (-1 sends all frames)", 1000);
+    static const FrameSimulatorOption<float> opt_interval("interval", "Transmission interval between frames (s)", 0.02);
+
+}
+
+#endif
```
Comment (MattTaylorDLS) [inline frameSimulator/include/FrameSimulatorOptions.h]: No new line at end of file

File: frameSimulator/include/FrameSimulatorPlugin.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    private:
+
+        /** Pointer to logger **/
+        LoggerPtr logger_;
+
+    };
+
+}
+
+#endif
```
Comment (MattTaylorDLS) [inline frameSimulator/include/FrameSimulatorPlugin.h]: No new line at end of file

File: frameSimulator/src/FrameSimulatorApp.cpp [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+
+
+    } catch (const std::exception &e) {
+        //LOG4CXX_ERROR(logger, "Caught unhandled exception in FrameSimulator, application will terminate: " << e.what());
+        throw;
+    }
+
+    return 0;
+
+}
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorApp.cpp]: No new line at end of file

File: frameSimulator/src/FrameSimulatorOption.cpp [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+          std::vector<std::string> tmp_list;
+          set_list_option(option_val, tmp_list);
+          list = tmp_list;
+      }
+
+}
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorOption.cpp]: No new line at end of file

File: frameSimulator/src/FrameSimulatorPlugin.cpp [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        opt_frames.get_val(vm, replay_numframes);
+        opt_interval.get_val(vm, replay_interval);
+
+        return true;
+
+    }
+
+}
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorPlugin.cpp]: No new line at end of file

File: frameSimulator/src/Packet.cpp [hunk 24]
```diff
@@ -0,0 +1,6 @@
+#include "Packet.h"
+
+namespace FrameSimulator {
+
+
+}
```
Comment (MattTaylorDLS) [inline frameSimulator/src/Packet.cpp]: No new line at end of file

File: frameSimulator/src/pcapFrameSimulatorPlugin.cpp [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+        replayer->prepare_packets(hdr, buffer);
+    }
+
+    void pcapFrameSimulatorPlugin::simulate() {
+
+        this->replay_frames();
+
+    }
+
+}
```
Comment (MattTaylorDLS) [inline frameSimulator/src/pcapFrameSimulatorPlugin.cpp]: No new line at end of file

File: frameSimulator/src/FrameSimulatorOption.cpp [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+     */
+    void set_list_option(const std::string &option_val, std::vector <std::string> &list) {
+        boost::split(list, option_val, boost::is_any_of(","), boost::token_compress_on);
+    }
+
+    /** Helper functions to parse comma separated optional command line arguments into lists
+     * /param[in] string of comma separated values
+     * /param[out] required output list as std::vector
+     */
+    void set_optionallist_option(const std::string &option_val, boost::optional <std::vector<std::string> > &list) {
```
Comment (ajgdls) [inline frameSimulator/src/FrameSimulatorOption.cpp]: To be able to compile this using DLS build program on RHEL6 I had to add a space between the two >> brackets.

Comment (ajgdls) [reply frameSimulator/src/FrameSimulatorOption.cpp]: Whoops @skyfrench I see you have already committed this change just now, and so this isn't some problem with compiler versions RHEL6/7.

**Comment by timcnicholls** [comment]
Also noticed that the FindPCAP.cmake file doesn't announce itself like the ZeroMQ, log4cxx, HDF ones do:
```
$ cmake ..
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   program_options
--   system
--   filesystem
--   unit_test_framework
--   date_time
--   thread

Looking for log4cxx headers and libraries
-- Root dir: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix
-- Include directories: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix/include/log4cxx
-- Libraries: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix/lib/liblog4cxx.so

Looking for ZeroMQ headers and libraries
-- Root dir: /aeg_sw/tools/CentOS7-x86_64/zeromq/4-2-3/prefix
-- Include directories: /usr/include
-- Libraries: /lib64/libzmq.so
-- Found PCAP: /usr/lib64/libpcap.so (Required is at least version "1.4.0") 
-- Performing Test PCAP_LINKS
-- Performing Test PCAP_LINKS - Success

Searching for HDF5
-- HDF5_ROOT set: /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix
Determining odin-data version
-- Git describe version: 0.6.1-22-g05581ef
-- major:0 minor:6 patch:1 sha1:g05581ef
-- short version: 0.6.1
-- HDF5 include files:  /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/include
-- HDF5 libs:           /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/lib/libhdf5.so/aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/lib/libhdf5_hl.so
-- HDF5 defs:           
-- Configuring done
-- Generating done
-- Build files have been written to: /u/tcn45/develop/projects/excalibur/odin-data/build
```
which makes it look like part of the ZeroMQ resolution step. Would it be possible to add a `Looking for ...` message call?

**Comment by skyfrench** [comment]
@timcnicholls suggested changes made :)

**Comment by skyfrench** [comment]
I've added --version to frameProcessor (and removed the version text from the help message) for consistency. The version reporting now matches for all three.

**Comment by timcnicholls** [comment]
Fantastic, thanks!
